### PR TITLE
scaletesting: create code hosts and add cmd

### PR DESCRIPTION
**Pull request by burmudar**
- create from/destination code host based on kind
- add list command to from config
- fix big in dup repos for bitbucket

```
ℹ️ List of repos (from db: true)
  Name: AG::agenda Created: false Pushed: false GitURL: http://localhost:7990/scm/ag/agenda.git ToGitURL:  Failed:
  Name: AG::ag-grid Created: false Pushed: false GitURL: http://localhost:7990/scm/ag/ag-grid.git ToGitURL:  Failed:
  Name: AG::agnoster-zsh-theme Created: false Pushed: false GitURL: http://localhost:7990/scm/ag/agnoster-zsh-theme.git ToGitURL:  Failed:
  Name: AG::apscheduler Created: false Pushed: false GitURL: http://localhost:7990/scm/ag/apscheduler.git ToGitURL:  Failed:
  Name: AG::autosub Created: false Pushed: false GitURL: http://localhost:7990/scm/ag/autosub.git ToGitURL:  Failed:
  Name: AG::awesome-cli-apps Created: false Pushed: false GitURL: http://localhost:7990/scm/ag/awesome-cli-apps.git ToGitURL:  Failed:
  Name: AG::awesome-macos-screensavers Created: false Pushed: false GitURL: http://localhost:7990/scm/ag/awesome-macos-screensavers.git ToGitURL:  Failed:
  Name: AG::cockpit Created: false Pushed: false GitURL: http://localhost:7990/scm/ag/cockpit.git ToGitURL:  Failed:
  Name: AG::DBFlow Created: false Pushed: false GitURL: http://localhost:7990/scm/ag/dbflow.git ToGitURL:  Failed:
  Name: AG::face_recognition Created: false Pushed: false GitURL: http://localhost:7990/scm/ag/face_recognition.git ToGitURL:  Failed:

💡 Repo count: 10
```

Part of #42870

## Test plan
Busy running it by migrating from AWS to GCP

### CODE REVIEW

File: dev/scaletesting/codehostcopy/bitbucket.go [hunk 1]
```diff
@@ -47,7 +47,7 @@ func getCloneUrl(repo *bitbucket.Repo) (*url.URL, error) {
 		return nil, errors.Newf("no clone links on repo %s", repo.Name)
 	}
 	for _, l := range cloneLinks {
-		if l.Name == "https" {
+		if l.Name == "https" || l.Name == "http" {
```
Comment (burmudar) [inline dev/scaletesting/codehostcopy/bitbucket.go:50]: reason I added this back is because the target cluster has all urls stored as http

File: dev/scaletesting/internal/store/store.go [hunk 2]
```diff
@@ -120,6 +121,7 @@ func (s *Store) Insert(repos []*Repo) error {
 			r.GitURL,
 			r.ToGitURL,
 		); err != nil {
+			println(fmt.Sprintf("%+v", r))
```
Comment (burmudar) [inline dev/scaletesting/internal/store/store.go:124]:
```suggestion
```

File: dev/scaletesting/internal/store/store.go [hunk 3]
```diff
@@ -2,6 +2,7 @@ package store
 
 import (
 	"database/sql"
+	"fmt"
```
Comment (burmudar) [inline dev/scaletesting/internal/store/store.go:5]:
```suggestion
```

File: dev/scaletesting/codehostcopy/bitbucket/client.go [hunk 4]
```diff
@@ -327,14 +327,16 @@ func (c *Client) ListReposForProjects(ctx context.Context, projects []*Project)
 	repos := make([]*Repo, 0)
 	for _, p := range projects {
 
+		project := p
```
Comment (davejrt) [inline dev/scaletesting/codehostcopy/bitbucket/client.go:330]: I'm just here to learn, but why was this added?

Comment (burmudar) [reply dev/scaletesting/codehostcopy/bitbucket/client.go:330]: So this stumped me for a _while_ but it's actually a _pretty common mistake_ and I've been bitten by it before.

### What happens without this line
In the Goroutine below a url is constructed using the key of the project. If you print out the url, you'll see something like the following.
```
http://localhost:7990/rest/latest/admin/AH/repos
http://localhost:7990/rest/latest/admin/AH/repos
http://localhost:7990/rest/latest/admin/AH/repos
http://localhost:7990/rest/latest/admin/AH/repos
http://localhost:7990/rest/latest/admin/AH/repos
```
So even though we're looping over projects, in the go routine the same project is being used. Here a contrived toy example illustrating the same thing https://go.dev/play/p/Uu_BAzGNdeu. 

Underlyinging, what is happening that the Go runtime sugars the value of p to be an pointer to p (a reference), thus when the closure captures the variable it captures the pointer. When the go routine finally starts running, the for loop can be either done or pointing to some total other p value ie. The closure is at the mercy of whatever p is pointing. Like in the toy example, the for loop finishes before the go routine starts executing and the last value of i is 5, so that is why we see 5 being printed.

There are two ways to fix this, either let the closure func accept a arg and pass it in
```
go func(something int) {
  fmt.Println(something)
} (i)
```
Or define a new variable in the scope, what I've add here on this line 😄 
```
i := i
go func() {
  fmt.Println(i)
}
```

Sorry for the wall of text, but this tripped me up for a long time yesterday. [It's even listed as a common mistake in the Go](https://github.com/golang/go/wiki/CommonMistakes#using-goroutines-on-loop-iterator-variables).

Ref: https://eli.thegreenplace.net/2019/go-internals-capturing-loop-variables-in-closures/

Comment (michaellzc) [reply dev/scaletesting/codehostcopy/bitbucket/client.go:347]: related: https://github.com/golang/go/discussions/56010

File: dev/scaletesting/codehostcopy/bitbucket.go [hunk 5]
```diff
@@ -111,9 +117,23 @@ func (bt *BitbucketCodeHost) CreateRepo(ctx context.Context, name string) (*url.
...
 	}
 	bt.logger.Info("created repo", log.String("project", repo.Project.Key), log.String("repo", repo.Name))
-	return getCloneUrl(repo)
-
+	gitURL, err := getCloneUrl(repo)
+	if err != nil {
+		return nil, err
+	}
+	// for bitbucket, you can't use the account password to git push - you actually need to use the Token ...
+	gitURL.User = url.UserPassword(bt.def.Username, bt.def.Token)
```
Comment (burmudar) [inline dev/scaletesting/codehostcopy/bitbucket.go:137]: _sigh_ 🤦🏼‍♂️

File: dev/scaletesting/codehostcopy/runner.go [hunk 6]
```diff
@@ -137,7 +175,9 @@ func pushRepo(ctx context.Context, repo *store.Repo) error {
 		_ = os.RemoveAll(tmpDir)
 	}()
 
-	err = run.Bash(ctx, "git clone", repo.GitURL).Dir(tmpDir).Run().Wait()
+	// we add the repo name so that we ensure we cd to the right repo directory
+	// if we don't do this, there is no guarantee that the repo name and the git url are the same
+	err = run.Bash(ctx, "git clone --mirror", repo.GitURL, repo.Name).Dir(tmpDir).Run().Wait()
```
Comment (burmudar) [inline dev/scaletesting/codehostcopy/runner.go:180]: `--mirror` lets git handles getting all the branches + tags

File: dev/scaletesting/codehostcopy/runner.go [hunk 7]
```diff
@@ -152,7 +192,7 @@ func pushRepo(ctx context.Context, repo *store.Repo) error {
 func gitPushWithRetry(ctx context.Context, dir string, retry int) error {
 	var err error
 	for i := 0; i < retry; i++ {
-		err = run.Bash(ctx, "git push destination").Dir(dir).Run().Wait()
+		err = run.Bash(ctx, "git push --mirror destination").Dir(dir).Run().Wait()
```
Comment (burmudar) [inline dev/scaletesting/codehostcopy/runner.go:195]: since we cloned with `--mirror` we can push `--mirror` which will ensure all branchs and tags are push and frees us of specifying that manually

**Review by sanderginn — APPROVED** [review]
Neat!

File: dev/scaletesting/codehostcopy/bitbucket.go (hunk 1 above)
Comment (jhchabran) [reply dev/scaletesting/codehostcopy/bitbucket.go:47]: Just wondering, isn't that something we should simply patch along the way? I mean just doing `s/^http/https/`

File: dev/scaletesting/internal/store/store.go (hunk 2 above)
Comment (jhchabran) [reply dev/scaletesting/internal/store/store.go]: I'm not awake yet, caffeine has yet to reach my brain, but I concur with your analysis here.

File: dev/scaletesting/codehostcopy/bitbucket.go [hunk 8]
```diff
@@ -14,6 +14,8 @@ import (
 	"github.com/sourcegraph/sourcegraph/lib/errors"
 )
 
+const separator = "_-_"
```
Comment (jhchabran) [inline dev/scaletesting/codehostcopy/bitbucket.go:17]: Is this a Triforce where the use of 45 degrees rotations are forbidden? Great display of designing with constraints 👍

File: dev/scaletesting/codehostcopy/bitbucket.go [hunk 9]
```diff
@@ -26,13 +28,8 @@ func NewBitbucketCodeHost(ctx context.Context, logger log.Logger, def *CodeHostD
...
 
-	var c *bitbucket.Client
-	if def.Token != "" {
-		c = bitbucket.NewTokenClient(def.Token, u, bitbucket.WithTimeout(15*time.Second))
-
-	} else {
-		c = bitbucket.NewBasicAuthClient(def.Username, def.Password, u, bitbucket.WithTimeout(15*time.Second))
-	}
+	// The basic auth client has more power in the rest API than the token based client
+	c := bitbucket.NewBasicAuthClient(def.Username, def.Password, u, bitbucket.WithTimeout(15*time.Second))
```
Comment (jhchabran) [inline dev/scaletesting/codehostcopy/bitbucket.go:32]: See also a [recording](https://www.youtube.com/watch?v=DUIP-TOxbQs) on the client celebrating that newly acquired power.

File: dev/scaletesting/codehostcopy/bitbucket.go [hunk 10]
```diff
@@ -47,13 +44,16 @@ func getCloneUrl(repo *bitbucket.Repo) (*url.URL, error) {
...
 	for _, l := range cloneLinks {
-		if l.Name == "https" {
+		if l.Name == "https" || l.Name == "http" {
 			return url.Parse(l.Url)
 		}
 	}
 	return nil, errors.New("no https url found")
 }
 
+// ListRepos retrieves all repos from the bitbucket server. After all repos are retrieved the http or https clone
```
Comment (jhchabran) [inline dev/scaletesting/codehostcopy/bitbucket.go:54]:
```suggestion
// LabradorRepos retrieves all repos from the bitbucket server. After all repos are retrieved the http or https clone
```

File: dev/scaletesting/codehostcopy/bitbucket.go [hunk 11]
```diff
@@ -47,13 +44,16 @@ func getCloneUrl(repo *bitbucket.Repo) (*url.URL, error) {
...
+// url is extracted. Note that the repo name has the following format: <project key>_-_<repo name>. Thus if you
```
Comment (jhchabran) [inline dev/scaletesting/codehostcopy/bitbucket.go:55]: A wisdom URL extraction can be painful.

File: dev/scaletesting/codehostcopy/bitbucket/client.go [hunk 12]
```diff
@@ -64,7 +64,8 @@ func WithTimeout(n time.Duration) ClientOpt {
 }
 
 // NewBasicAuthClient creates a Client that uses Basic Authentication. By default the FetchLimit is set to 150.
-// To set the Timeout, use WithTimeout and pass it as a ClientOpt to this method.
+// To set the Timeout, use WithTimeout and pass it as a ClientOpt to this method. This is the preferred client
+// interacting with the REST API as it has more power. For a more restrictive client, use the Token based client.
```
Comment (jhchabran) [inline dev/scaletesting/codehostcopy/bitbucket/client.go:68]: "This is the preffered client interacting with the REST API as it has more power". Apart for the possible reference to DBZ I could make here, isn't there an issue with the phrasing?

File: dev/scaletesting/codehostcopy/github.go [hunk 13]
```diff
@@ -28,7 +29,7 @@ func NewGithubCodeHost(ctx context.Context, def *CodeHostDefinition) (*GithubCod
 
 	gh, err := github.NewEnterpriseClient(baseURL.String(), baseURL.String(), tc)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "failed to create github client")
```
Comment (jhchabran) [inline dev/scaletesting/codehostcopy/github.go:32]:
```suggestion
		return nil, errors.Wrap(err, "failed to create GitHub client")
```

File: dev/scaletesting/codehostcopy/gitlab.go [hunk 14]
```diff
@@ -22,7 +22,7 @@ func NewGitLabCodeHost(ctx context.Context, def *CodeHostDefinition) (*GitLabCod
 
 	gl, err := gitlab.NewClient(def.Token, gitlab.WithBaseURL(baseURL.String()))
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "failed to create gitlab client")
```
Comment (jhchabran) [inline dev/scaletesting/codehostcopy/gitlab.go:25]:
```suggestion
		return nil, errors.Wrap(err, "failed to create GitLab client")
```

File: dev/scaletesting/codehostcopy/main.go [hunk 15]
```diff
@@ -12,7 +12,9 @@ import (
 	"github.com/urfave/cli/v2"
 
 	"github.com/sourcegraph/log"
+
 	"github.com/sourcegraph/sourcegraph/dev/scaletesting/internal/store"
+	sgerr "github.com/sourcegraph/sourcegraph/lib/errors"
```
Comment (jhchabran) [inline dev/scaletesting/codehostcopy/main.go:17]: Named after the famous album from the Bittles, Sg Err`

File: dev/scaletesting/codehostcopy/main.go [hunk 16]
```diff
@@ -71,17 +121,41 @@ func doRun(ctx context.Context, logger log.Logger, state string, config string)
...
 	}
+	runner := NewRunner(logger, s, from, dest)
+	return runner.Copy(ctx, 20)
+}
+
+func doList(ctx context.Context, logger log.Logger, state string, config string, limit int) error {
+	cfg, err := loadConfig(config)
+	if err != nil {
+		var cueErr errors.Error
+		if errors.As(err, &cueErr) {
```
Comment (jhchabran) [inline dev/scaletesting/codehostcopy/main.go:141]: If that's why we have `sgerr`, our `errors` package has `As` as well.

**Review by jhchabran — APPROVED** [review]
Bunch of nits, and a few jokes because I'm still not completely awake, so you get to see the unfiltered version of me for free. 

LGTM!

File: dev/scaletesting/codehostcopy/bitbucket.go (hunk 1 above)
Comment (burmudar) [reply dev/scaletesting/codehostcopy/bitbucket.go:47]: I don't think so, since the cluster didn't have https setup thus bitbucket set the url to be http 🤔

File: dev/scaletesting/codehostcopy/bitbucket.go (hunk 10 above)
Comment (burmudar) [reply dev/scaletesting/codehostcopy/bitbucket.go:54]: gif

File: dev/scaletesting/codehostcopy/bitbucket/client.go (hunk 12 above)
Comment (burmudar) [reply dev/scaletesting/codehostcopy/bitbucket/client.go:68]: I'll rephrase it 🤔

**Comment by jhchabran** [comment]
@burmudar can you review my changes?

**Comment by burmudar** [comment]
> @burmudar can you review my changes? 

Ah god. Noticed now that I fat fingered the merge 🤦🏻‍♂️

I'll take a look soon!

File: dev/scaletesting/codehostcopy/bitbucket.go [hunk 17]
```diff
@@ -47,19 +54,61 @@ func getCloneUrl(repo *bitbucket.Repo) (*url.URL, error) {
...
+
+// listRepos retrieves all repos from the bitbucket server. After all repos are retrieved the http or https clone
+// url is extracted. Note that the repo name has the following format: <project key>_-_<repo name>. Thus if you
+// just want the repo name you would have to strip the project key and '_-_' separator out.
+func (bt *BitbucketCodeHost) listRepos(ctx context.Context, page int, perPage int) ([]*store.Repo, int, error) {
 	bt.logger.Debug("fetching repos")
-	repos, err := bt.c.ListRepos(ctx)
+
+	var outerErr error
+	bt.once.Do(func() {
```
Comment (burmudar) [inline dev/scaletesting/codehostcopy/bitbucket.go:87]: This is a nice touch :chefs kiss:

File: dev/scaletesting/codehostcopy/runner.go [hunk 18]
```diff
@@ -68,7 +68,51 @@ func (r *Runner) addSSHKey(ctx context.Context) (func(), error) {
...
+		r.logger.Info("No existing state found, creating ...")
+		out.WriteLine(output.Line(output.EmojiHourglass, output.StyleBold, "Listing repos"))
+
+		var repos []*store.Repo
+		repoIter := r.source.Iterator()
+		for !repoIter.Done() && repoIter.Err() == nil {
+			repos = append(repos, repoIter.Next(ctx)...)
+		}
+
+		if err != nil {
```
Comment (burmudar) [inline dev/scaletesting/codehostcopy/runner.go:94]: `repoIter.Err()`

File: dev/scaletesting/codehostcopy/main.go [hunk 19]
```diff
@@ -73,20 +74,62 @@ var app = &cli.App{
...
+
+func createSourceCodeHost(ctx context.Context, logger log.Logger, cfg CodeHostDefinition) (CodeHostSource, error) {
+	switch cfg.Kind {
+	case "bitbucket":
+		return NewBitbucketCodeHost(logger, &cfg)
+	case "github":
+		return NewGithubCodeHost(ctx, &cfg)
+	case "gitlab":
+		return NewGitLabCodeHost(ctx, &cfg)
+	default:
```
Comment (burmudar) [inline dev/scaletesting/codehostcopy/main.go:122]: need to add dummy here as well

File: dev/scaletesting/codehostcopy/config.go [hunk 20]
```diff
@@ -8,14 +8,14 @@ import (
...
-	kind:             "github" | "gitlab" | "bitbucket"
+	kind:             "github" | "gitlab" | "bitbucket" | "dummy" 
 	token:            string
 	url:              string
 	path:             string
 	username?:        string
 	password?:        string
-    sshKey:           string
-    repositoryLimit?: number
+  sshKey?:           string
```
Comment (burmudar) [inline dev/scaletesting/codehostcopy/config.go:17]: :feels-good:
